MoreLinq/CountBy.cs

```diff
@@ -94,17 +94,17 @@ bool TryGetIndex(TKey key, out int i)
 
                     keys = new List<TKey>();
                     counts = new List<int>();
-                    var havePrevKey = false;
-                    var prevKey = default(TKey);
+                    var prevKey = (false, default(TKey));
                     var index = 0;
 
                     foreach (var item in source)
                     {
                         var key = keySelector(item);
 
                         if (// key same as the previous? then re-use the index
-                            havePrevKey && cmp.GetHashCode(prevKey) == cmp.GetHashCode(key)
-                                         && cmp.Equals(prevKey, key)
+                            prevKey is (true, {} pk)
```

Comment:
I feel it belongs in this PR because pattern matching here avoids using dammit (`!`) and…

> This is a hot path and changing simple calls for pattern matching could reduce performance.

…the cost seems to be [negligible](https://gist.github.com/atifaziz/b1fcf40a266ae9fecff0b33a77c5e06b) ([see also on SharpLab](https://sharplab.io/#gist:b1fcf40a266ae9fecff0b33a77c5e06b)) as the following test:

```c#
using System;

public class C {
    public void M((bool, object) x) {
        if (x is (true, {} obj)) {
            Console.WriteLine(obj);
        }
    }
    public void N(bool f, object obj) {
        if (f) {
            Console.WriteLine(obj);
        }
    }
}
```

shows it adds just 6 bytes of x64 instructions (assuming that's a satisfactory measure):

```asm
; Core CLR v4.700.20.26901 on amd64

C..ctor()
    L0000: ret

C.M(System.ValueTuple`2<Boolean,System.Object>)
    L0000: cmp byte ptr [rdx+8], 0
    L0004: je short L001b
    L0006: mov rcx, [rdx]
    L0009: test rcx, rcx
    L000c: je short L001b
    L000e: mov rax, System.Console.WriteLine(System.Object)
    L0018: jmp rax
    L001b: ret

C.N(Boolean, System.Object)
    L0000: test dl, dl
    L0002: je short L0014
    L0004: mov rcx, r8
    L0007: mov rax, System.Console.WriteLine(System.Object)
    L0011: jmp rax
    L0014: ret
```

If you still feel strongly about this going into a separate PR then perhaps you open that as a new issue and we revert via a PR. What's your final verdict?

Reply:
OK, it seems there is not much difference:

    public class Benchmarks
    {
        private readonly (bool, object)[] data;

        public Benchmarks() {
            data = new [] {
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
            };
        }
        
        [Benchmark]
        public void MTest()
        {
            foreach (var d in data)
                M(d);
        }
        [Benchmark]
        public void NTest()
        {
            foreach (var d in data)
                N(d.Item1, d.Item2);
        }

        public void M((bool, object) x) {
            if (x is (true, {} obj)) {
                obj.GetHashCode();
            }
        }
        public void N(bool f, object obj) {
            if (f) {
                obj.GetHashCode();
            }
        }
    }


```
// * Summary *

BenchmarkDotNet=v0.11.5, OS=debian 
Intel Core i7-7600U CPU 2.80GHz (Kaby Lake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.1.302
  [Host]     : .NET Core 3.0.3 (CoreCLR 4.700.20.6603, CoreFX 4.700.20.6701), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.3 (CoreCLR 4.700.20.6603, CoreFX 4.700.20.6701), 64bit RyuJIT


| Method |     Mean |     Error |    StdDev |
|------- |---------:|----------:|----------:|
|  MTest | 134.4 ns | 0.4219 ns | 0.3947 ns |
|  NTest | 136.0 ns | 0.2768 ns | 0.2312 ns |
```

Let's merge as it is